File: internal/parser/hcl.go
// Package parser wraps hashicorp/hcl (the parser Terraform itself uses) to
// turn raw .tf source into a normalized slice of Resource values that the
// rule engine can inspect without touching the HCL AST directly.
package parser

import (
	"fmt"

	"github.com/hashicorp/hcl/v2"
	"github.com/hashicorp/hcl/v2/hclsyntax"
	"github.com/zclconf/go-cty/cty"
)

// ParseFile parses a single .tf file's contents and returns the blocks it
// declares — resources, module calls and data sources. Diagnostics from
// malformed HCL are returned as an error; callers should treat parse failures
// as a finding of their own rather than crashing the whole scan.
func ParseFile(filename string, src []byte) ([]*Resource, error) {
	return ParseFileWithContext(filename, src, nil)
}

// ParseFileWithContext is ParseFile with a scope for resolving references.
//
// Without a context, `password = var.db_password` is simply unresolvable and
// every rule skips it. With one built from the surrounding directory's
// `locals` blocks and `variable` defaults (see BuildScope), that same line
// resolves to the value it will actually carry — which is how a password left
// in a variable default gets caught instead of hiding one indirection away.
//
// A reference the scope can't resolve stays unresolved rather than guessed,
// so a richer scope only ever finds more, never differently.
func ParseFileWithContext(filename string, src []byte, ctx *hcl.EvalContext) ([]*Resource, error) {
	body, err := parseBody(filename, src)
	if err != nil {
		return nil, err
	}

	var resources []*Resource
	for _, block := range body.Blocks {
		switch {
		case block.Type == "resource" && len(block.Labels) == 2:
			resources = append(resources, blockToResource(filename, block, KindResource, block.Labels[0], block.Labels[1], ctx))
		case block.Type == "data" && len(block.Labels) == 2:
			resources = append(resources, blockToResource(filename, block, KindData, block.Labels[0], block.Labels[1], ctx))
		case block.Type == "module" && len(block.Labels) == 1:
			// A module call has no type of its own; "module" stands in so
			// schema-driven rules, which look types up in a provider pack,
			// find nothing and skip it.
			resources = append(resources, blockToResource(filename, block, KindModule, "module", block.Labels[0], ctx))
		}
	}
	return resources, nil
}

func parseBody(filename string, src []byte) (*hclsyntax.Body, error) {
	file, diags := hclsyntax.ParseConfig(src, filename, hcl.InitialPos)
	if diags.HasErrors() {
		return nil, fmt.Errorf("hcl parse error in %s: %s", filename, diags.Error())
	}
	body, ok := file.Body.(*hclsyntax.Body)
	if !ok {
		return nil, fmt.Errorf("unexpected body type for %s", filename)
	}
	return body, nil
}

func blockToResource(filename string, block *hclsyntax.Block, kind Kind, typeName, name string, ctx *hcl.EvalContext) *Resource {
	r := &Resource{
		Kind:       kind,
		Type:       typeName,
		Name:       name,
		File:       filename,
		DefRange:   block.DefRange(),
		Attributes: map[string]*Attribute{},
	}

	for attrName, attr := range block.Body.Attributes {
		r.Attributes[attrName] = attrToAttribute(attrName, attr, ctx)
	}

	for _, nested := range block.Body.Blocks {
		if nested.Type == "lifecycle" {
			r.HasLifecycleBlock = true
			r.LifecycleRange = nested.DefRange()
			if pdAttr, ok := nested.Body.Attributes["prevent_destroy"]; ok {
				r.PreventDestroyRange = pdAttr.SrcRange
				if v, diags := pdAttr.Expr.Value(ctx); !diags.HasErrors() && v.Type() == cty.Bool {
					b := v.True()
					r.PreventDestroyValue = &b
				}
			}
			continue
		}
		nb := &NestedBlock{
			Type:       nested.Type,
			Labels:     nested.Labels,
			Range:      nested.DefRange(),
			Attributes: map[string]*Attribute{},
		}
		for attrName, attr := range nested.Body.Attributes {
			nb.Attributes[attrName] = attrToAttribute(attrName, attr, ctx)
		}
		r.Blocks = append(r.Blocks, nb)
	}

	return r
}

func attrToAttribute(name string, attr *hclsyntax.Attribute, ctx *hcl.EvalContext) *Attribute {
	a := &Attribute{
		Name:  name,
		Range: attr.SrcRange,
	}

	// Try the expression on its own first. If that works the value was written
	// inline, and there is no indirection worth reporting.
	if v, diags := attr.Expr.Value(nil); !diags.HasErrors() {
		a.IsLiteral = true
		a.RawValue = ctyValueToString(v)
		return a
	}

	if ctx == nil {
		// Expression references a variable/resource/function we can't resolve
		// (no plan, no state). Leave RawValue empty; rules that need a literal
		// value will simply skip this attribute.
		return a
	}

	v, diags := attr.Expr.Value(ctx)
	if diags.HasErrors() || !v.IsWhollyKnown() {
		return a
	}
	a.IsLiteral = true
	a.RawValue = ctyValueToString(v)
	a.ResolvedFrom = firstTraversalName(attr.Expr)
	return a
}

// firstTraversalName renders the reference an expression reads from, e.g.
// "var.db_password" or "local.admin_pw", for use in a finding message.
func firstTraversalName(expr hclsyntax.Expression) string {
	for _, traversal := range expr.Variables() {
		name := ""
		for i, step := range traversal {
			switch t := step.(type) {
			case hcl.TraverseRoot:
				name = t.Name
			case hcl.TraverseAttr:
				name += "." + t.Name
			default:
				// An index or splat: the root is still the useful part.
			}
			if i > 1 {
				break
			}
		}
		if name != "" {
			return name
		}
	}
	return ""
}

// ctyValueToString renders a literal cty.Value as plain text for pattern
// matching (e.g. comparing ForceNew attribute values across revisions, or
// regex-matching against "0.0.0.0/0").
func ctyValueToString(v cty.Value) string {
	if v.IsNull() {
		return ""
	}
	switch {
	case v.Type() == cty.String:
		return v.AsString()
	case v.Type() == cty.Bool:
		if v.True() {
			return "true"
		}
		return "false"
	case v.Type() == cty.Number:
		return v.AsBigFloat().String()
	case v.Type().IsListType(), v.Type().IsTupleType(), v.Type().IsSetType():
		out := ""
		it := v.ElementIterator()
		for it.Next() {
			_, ev := it.Element()
			if out != "" {
				out += ","
			}
			out += ctyValueToString(ev)
		}
		return out
	default:
		return ""
	}
}

File: src/tfpdf/parser/hcl.py
"""Transformation de source .tf brute en le modèle Resource normalisé.

Port de internal/parser/hcl.go.
"""

from __future__ import annotations

from .. import hcl
from ..hcl import Block, EvalContext, HCLParseError, Value
from ..hcl.ast import Expression
from .model import Attribute, Kind, NestedBlock, Resource


def parse_file(filename: str, source: bytes) -> list[Resource]:
    """Analyse un fichier .tf et rend les blocs qu'il déclare.

    Lève `HCLParseError` sur du HCL malformé ; les appelants doivent traiter un
    échec d'analyse comme une découverte à eux plutôt que de faire tomber tout
    le scan.
    """
    return parse_file_with_context(filename, source, None)


def parse_file_with_context(
    filename: str, source: bytes, context: EvalContext | None
) -> list[Resource]:
    """`parse_file`, avec une portée pour résoudre les références.

    Sans contexte, `password = var.db_password` est simplement irrésoluble et
    chaque règle le saute. Avec un contexte construit depuis les blocs `locals`
    et les valeurs par défaut de `variable` du répertoire environnant (voir
    `build_scope`), cette même ligne se résout en la valeur qu'elle portera
    réellement — c'est ainsi qu'un mot de passe laissé dans une valeur par
    défaut est attrapé au lieu de se cacher à une indirection de distance.

    Une référence que la portée ne peut pas résoudre reste non résolue plutôt
    que devinée, si bien qu'une portée plus riche ne trouve jamais que davantage,
    jamais autre chose.
    """
    body = _parse_body(filename, source)

    resources: list[Resource] = []
    for block in body.blocks:
        if block.type == "resource" and len(block.labels) == 2:
            resources.append(
                _block_to_resource(
                    filename, block, Kind.RESOURCE, block.labels[0], block.labels[1], context
                )
            )
        elif block.type == "data" and len(block.labels) == 2:
            resources.append(
                _block_to_resource(
                    filename, block, Kind.DATA, block.labels[0], block.labels[1], context
                )
            )
        elif block.type == "module" and len(block.labels) == 1:
            # A module call has no type of its own; "module" stands in so
            # schema-driven rules, which look types up in a provider pack,
            # find nothing and skip it.
            resources.append(
                _block_to_resource(filename, block, Kind.MODULE, "module", block.labels[0], context)
            )
    return resources


def _parse_body(filename: str, source: bytes) -> hcl.Body:
    file, diagnostics = hcl.parse_config(source, filename)
    if diagnostics.has_errors():
        raise HCLParseError(diagnostics)
    return file.body


def _block_to_resource(
    filename: str,
    block: Block,
    kind: Kind,
    type_name: str,
    name: str,
    context: EvalContext | None,
) -> Resource:
    resource = Resource(
        kind=kind, type=type_name, name=name, file=filename, def_range=block.def_range()
    )

    for attr_name, attribute in block.body.attributes.items():
        resource.attributes[attr_name] = _attr_to_attribute(attr_name, attribute, context)

    for nested in block.body.blocks:
        if nested.type == "lifecycle":
            resource.has_lifecycle_block = True
            resource.lifecycle_range = nested.def_range()
            prevent_destroy_attribute = nested.body.attributes.get("prevent_destroy")
            if prevent_destroy_attribute is not None:
                resource.prevent_destroy_range = prevent_destroy_attribute.src_range
                value, diagnostics = prevent_destroy_attribute.expr.value(context)
                if not diagnostics.has_errors() and value.type is hcl.BOOL and not value.is_null():
                    resource.prevent_destroy_value = value.true()
            continue
        nested_block = NestedBlock(
            type=nested.type, labels=list(nested.labels), range=nested.def_range()
        )
        for attr_name, attribute in nested.body.attributes.items():
            nested_block.attributes[attr_name] = _attr_to_attribute(attr_name, attribute, context)
        resource.blocks.append(nested_block)

    return resource


def _attr_to_attribute(
    name: str, attribute: hcl.Attribute, context: EvalContext | None
) -> Attribute:
    parsed_attribute = Attribute(name=name, range=attribute.src_range)

    # Try the expression on its own first. If that works, the value was written
    # inline and there is no indirection worth reporting.
    value, diagnostics = attribute.expr.value(None)
    if not diagnostics.has_errors():
        parsed_attribute.is_literal = True
        parsed_attribute.raw_value = cty_value_to_string(value)
        return parsed_attribute

    if context is None:
        # References a variable/resource/function we cannot resolve (no plan,
        # no state). Leave raw_value empty; rules that need a literal simply
        # skip this attribute.
        return parsed_attribute

    value, diagnostics = attribute.expr.value(context)
    if diagnostics.has_errors() or not value.is_wholly_known():
        return parsed_attribute
    parsed_attribute.is_literal = True
    parsed_attribute.raw_value = cty_value_to_string(value)
    parsed_attribute.resolved_from = first_traversal_name(attribute.expr)
    return parsed_attribute


def first_traversal_name(expr: Expression) -> str:
    """Rend la référence dont une expression lit sa valeur —
    « var.db_password », « local.admin_pw » — pour usage dans un message de
    découverte."""
    for traversal in expr.variables():
        name = traversal.render(max_steps=2)
        if name:
            return name
    return ""


def cty_value_to_string(v: Value) -> str:
    """Rend une valeur littérale en texte brut, pour la comparaison de motifs :
    comparer des valeurs d'attributs ForceNew d'une révision à l'autre, ou
    chercher « 0.0.0.0/0 » par expression régulière.

    **Défaut connu, porté tel quel plutôt que corrigé ici.** Un objet ou un map
    rend `""` — seuls les chaînes, nombres, booléens et séquences sont traités.
    Une règle personnalisée « doit avoir des tags » écrite ainsi :

        attribute: tags
        pattern: ".+"
        negate: true

    voit donc `""` pour `tags = { env = "prod" }`, conclut que le motif n'a pas
    correspondu, et **se déclenche sur une ressource qui définit bien des
    tags**. C'est un faux positif qui bloque une PR chez un client.

    Le comportement a été confirmé identique côté Go — mêmes deux découvertes
    sur les mêmes deux lignes — et
    `test_customrules.py::test_negated_rule_misfires_on_an_object_valued_attribute`
    l'épingle, pour que le port reste fidèle et qu'une correction apparaisse
    comme un changement délibéré des deux côtés à la fois. L'endroit naturel
    pour corriger est ici : rendre un objet comme ses paires `clé=valeur`
    triées.
    """
    if v.is_null() or v.is_unknown():
        return ""
    if v.type is hcl.STRING:
        return v.as_string()
    if v.type is hcl.BOOL:
        return "true" if v.true() else "false"
    if v.type is hcl.NUMBER:
        return v.as_number_string()
    if v.type.is_list_type() or v.type.is_tuple_type() or v.type.is_set_type():
        return ",".join(
            cty_value_to_string(element_value) for _, element_value in v.element_iterator()
        )
    return ""
